### PR TITLE
V62-Addons - 32 Bit saturating add, 32 Bit saturating add for double …

### PR DESCRIPTION
**Pull request by pranavb-ca**
…vectors, narrowing u32 to u16, broadcast instruction for half-words and bytes

### CODE REVIEW

**Comment by pranavb-ca** [comment]
@ronlieb @dpalermo : PTAL

**Comment by dpalermo** [comment]
Any way we can get Travis builders working with the cloned repo?  ...or is that something we give up with the clone?

File: src/CodeGen_Hexagon.cpp [hunk 1]
```diff
@@ -40,6 +40,7 @@ using namespace llvm;
 #endif
 
 #define IPICK(is_128B, i64) (is_128B ? i64##_128B : i64)
+#define V62PICK(is_128B, is_v62, i64, v62_i64) (is_v62 ? IPICK(is_128B, v62_i64) : IPICK(is_128B, i64))
```
Comment (dpalermo) [inline src/CodeGen_Hexagon.cpp]: I'm wondering if it would be better to not create a macro for specifically selecting the intrinsic given V62 and vector width as it would be better to expose the selection logic so it is more understandable and controllable (esp. as more architectures and vector width combinations are added).

For example, with this change, the broadcast is selected as:
 
 ```
        // Broadcasts
+        { V62PICK(is_128B, is_v62, Intrinsic::hexagon_V6_lvsplatw, Intrinsic::hexagon_V6_lvsplatb), u8v1,  "splat.b", {u8} },
+        { V62PICK(is_128B, is_v62, Intrinsic::hexagon_V6_lvsplatw, Intrinsic::hexagon_V6_lvsplath), u16v1,  "splat.h", {u16} },
         { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u32v1,  "splat.w", {u32} },
```

but it could be more clearly selected without the macro as:

 ```
        // Broadcasts
        { is_v62 ?
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatb) :
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u8v1,  "splat.b", {u8} },
        { is_v62 ?
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplath) :
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u16v1,  "splat.h", {u16} },
        { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u32v1, "splat.w", {u32} }, 
```

...or if people really wanted a macro instead of the ternary op, maybe just define a generic TGTPICK macro that doesn't specify V62 in the name (as it could be used for any feature) and doesn't include the vector width:

#define TGTPICK(feature, has, has_not) (feature ? has : has_not)

 ```
        // Broadcasts
        { TGTPICK(is_v62,
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatb),
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw)), u8v1,  "splat.b", {u8} },
        { TGTPICK(is_v62,
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplath),
            IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw)), u16v1,  "splat.h", {u16} },
        { IPICK(is_128B, Intrinsic::hexagon_V6_lvsplatw), u32v1, "splat.w", {u32} }, 
```

Comment (pranavb-ca) [reply src/CodeGen_Hexagon.cpp]: +1 for the TGPICK approach

File: test/correctness/simd_op_check.cpp [hunk 2]
```diff
@@ -1493,7 +1495,12 @@ struct Test {
 
         check("vadd(v*.b,v*.b)", hvx_width/1, u8_1 + u8_2);
         check("vadd(v*.h,v*.h)", hvx_width/2, u16_1 + u16_2);
-        check("vadd(v*.w,v*.w)", hvx_width/4, u32_1 + u32_2);
+        if (is_v62) {
+            // v62 - 32 bit saturating add
+            check("vadd(v*.uw,v*.uw):sat", hvx_width/4, u32_1 + u32_2);
```
Comment (pranavb-ca) [inline test/correctness/simd_op_check.cpp]: I dont think this is the right pattern to match. u32_1 + u32_2 is adding two unsigned 32 bit values. Unsigned overflow is defined (modulo). saturated addition clamps the add to the UINT_MAX instead. You should really be matching this pattern u32_sat(u64(u32_1) + u64(u32_1));

File: src/CodeGen_Hexagon.cpp [hunk 3]
```diff
@@ -293,6 +295,9 @@ void CodeGen_Hexagon::init_module() {
         { IPICK(is_128B, Intrinsic::hexagon_V6_vsathub),  u8v1,  "trunc_satub.vh",  {i16v2} },
         { IPICK(is_128B, Intrinsic::hexagon_V6_vsatwh),   i16v1, "trunc_sath.vw",   {i32v2} },
 
+        // v62 - Saturating narrowing cast
+        { TGTPICK(is_v62, Intrinsic::hexagon_V6_vsatuwuh, Intrinsic::hexagon_V6_vsatwh),   u16v1, "trunc_satuh.vuw",   {u32v2} },
```
Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: This might break v60 - A large unsigned number will be treated as negative by hexagon_V6_vsatwh and will be saturated to 0 instead of UINT16_MAX.

File: src/HexagonOptimize.cpp [hunk 4]
```diff
@@ -767,6 +767,9 @@ class OptimizePatterns : public IRMutator {
             { "halide.hexagon.trunc_satub_shr.vh.h", u8_sat(wild_i16x/wild_i16), Pattern::DeinterleaveOp0 | Pattern::ExactLog2Op1 },
             { "halide.hexagon.trunc_satuh_shr.vw.w", u16_sat(wild_i32x/wild_i32), Pattern::DeinterleaveOp0 | Pattern::ExactLog2Op1 },
             { "halide.hexagon.trunc_sath_shr.vw.w",  i16_sat(wild_i32x/wild_i32), Pattern::DeinterleaveOp0 | Pattern::ExactLog2Op1 },
+            
+            // v62 - Saturating narrowing cast
+            { "halide.hexagon.trunc_satuh.vuw", u16_sat(wild_u32x) },
```
Comment (pranavb-ca) [inline src/HexagonOptimize.cpp]: See comment above about large unsigned values.

File: test/correctness/simd_op_check.cpp (hunk 2 above)
Comment (pranavb-ca) [reply test/correctness/simd_op_check.cpp]: Again, there is no saturation in the above addition. For example, saturating a 16 bit value, say 'a', to an 8 bit unsigned value is equivalent to a_sat = clamp(a, 0, 255) i.e. a_sat = min(max(0, a), 255). So a saturating add has to saturate the result of the add.

File: src/CodeGen_Hexagon.cpp [hunk 5]
```diff
@@ -1331,9 +1342,17 @@ int CodeGen_Hexagon::native_vector_bits() const {
 
 void CodeGen_Hexagon::visit(const Add *op) {
     if (op->type.is_vector()) {
-        value = call_intrin(op->type,
-                            "halide.hexagon.add" + type_suffix(op->a, op->b, false),
-                            {op->a, op->b});
+        if (op->a.type().is_uint() && op->b.type().is_uint() && op->a.type().bits()==32 && op->b.type().bits()==32) {
```
Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: Please see comment about saturating adds below.

**Review by pranavb-ca — COMMENTED** [review]
@aankit-ca - Please review my comments.

File: src/CodeGen_Hexagon.cpp [hunk 6]
```diff
@@ -348,7 +349,7 @@ void CodeGen_Hexagon::init_module() {
         { IPICK(is_128B, Intrinsic::hexagon_V6_vadduhsat_dv), u16v2, "satuh_add.vuh.vuh.dv", {u16v2, u16v2} },
         { IPICK(is_128B, Intrinsic::hexagon_V6_vaddhsat_dv),  i16v2, "sath_add.vh.vh.dv",    {i16v2, i16v2} },
         { IPICK(is_128B, Intrinsic::hexagon_V6_vaddwsat_dv),  i32v2, "satw_add.vw.vw.dv",    {i32v2, i32v2} },
-
+
```
Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: Whitespace. Fix.

File: src/HexagonOptimize.cpp [hunk 7]
```diff
@@ -799,7 +802,7 @@ class OptimizePatterns : public IRMutator {
             { "halide.hexagon.pack.vh", u8(wild_i16x) },
             { "halide.hexagon.pack.vh", i8(wild_u16x) },
             { "halide.hexagon.pack.vh", i8(wild_i16x) },
-            { "halide.hexagon.pack.vw", u16(wild_u32x) },
+            // { "halide.hexagon.pack.vw", u16(wild_u32x) },
```
Comment (pranavb-ca) [inline src/HexagonOptimize.cpp]: Did you mean to leave this code commented?

File: test/correctness/simd_op_check.cpp [hunk 8]
```diff
@@ -1247,7 +1247,7 @@ struct Test {
             check(arm32 ? "vshll.u16" : "ushll", 4*w, u32(u16_1)*16);
             check(arm32 ? "vshll.u32" : "ushll", 2*w, u64(u32_1)*16);
 
-            // VSHR     I	-	Shift Right
+            // VSHR     I   -   Shift Right
```
Comment (pranavb-ca) [inline test/correctness/simd_op_check.cpp]: These changes are not part of the fix. They will simply corrupt the git blame history. Please undo these changes.

**Comment by aankit-ca** [comment]
@pranavb-ca @dpalermo @ronlieb I have corrected the whitespaces problem. Please review the latest push.

File: test/correctness/simd_op_check.cpp [hunk 9]
```diff
@@ -1755,8 +1769,14 @@ struct Test {
         check("vnot(v*)", hvx_width/2, ~u16_1);
         check("vnot(v*)", hvx_width/4, ~u32_1);
 
-        check("vsplat(r*)", hvx_width/1, in_u8(0));
-        check("vsplat(r*)", hvx_width/2, in_u16(0));
+        if (is_v62) {
+            // v62 - Broadcasting unsigned 8 bit and 16 bit scalers
+            // check("v*.b = vsplat(r*)", hvx_width/1, in_u8(0));
+            // check("v*.h = vsplat(r*)", hvx_width/2, in_u16(0));
```
Comment (dpalermo) [inline test/correctness/simd_op_check.cpp]: Should these be commented out?

Comment (aankit-ca) [reply test/correctness/simd_op_check.cpp]: This is the vsplat problem I was facing. Actually this instruction should ideally be generated, but presently the instruction doesn't actually generate even though there is a call to use the intrinsic that generates this instruction. What should I do about this for the time being?

Comment (pranavb-ca) [reply test/correctness/simd_op_check.cpp]: @aankit-ca - For the time being, get rid of the if condition, essentially keeping the test as it. That means, vectorizing in_u8(0) should lead to a regular "vsplat" even on v62. This is to say that whenever we don't have v62 support we fall back to v60.

Comment (aankit-ca) [reply test/correctness/simd_op_check.cpp]: Ok Pranav. Thanks
